### ISSUE TITLE
Memory Leak Issue

### ISSUE THREAD

**Issue by ngxbac** [catalyst-team/catalyst#15]
I just want to inform that I am countering with Memory leak issue. 
I am not 100% sure that It comes from catalyst. But I am sure that when I use my own code written by myself (without catalyst), there is no leaking issue. I tried to pre-procedure this issue from your [examples](https://github.com/Scitator/catalyst-examples), but it runs very fast and we could not see the leak.

Following is my leak descriptions:

- Image classification task with ~17M images
- The memory leak occurs every epoch and it happens at the next epoch again.
EX:  
At the beginning of epoch 0, my memory is:

  - RAM: 13G/16G
  - Swap: 0/50G 

  At the end of epoch 0, my memory is:
  - RAM: 16G/16G
  - Swap: 25G/50G

  When epoch 0 ends, the memories (RAM + Swap) are released, (RAM: 13G/16G, SwapL: 0/50G). The same phenomena occurs at the next epoch 1.

I am debugging to see where the issue comes from. Could you please double check with your bigger data?

**Comment by ngxbac** [comment]
Updated:

1. Train with bigger data by using [finetune](https://github.com/Scitator/catalyst-examples/tree/master/finetune)  
I downloaded [dogscats](http://files.fast.ai/data/dogscats.zip) dataset from fastai and tried to run finetune from your examples. There is no leak issue during training one epoch
It seems that the leak comes from my code

2. Debug my code
I am able to figure out where the leak comes from. It occurs at my custom dataset.
```python
    def __getitem__(self, idx):

        # Get token
        token = self.tokens[idx]

        # Get drawing
        drawing = self.drawings[idx]
        stroke = stroke_data(drawing).astype(np.float32)
        # stroke = np.expand_dims(stroke, axis=0)
        stroke = np.transpose(stroke, (1, 0)).astype(np.float32)

        drawing = eval(drawing)
        label = self.labels[idx]

        # Plot the image
        img, img_gray = drawing_to_image(drawing, IMG_SIZE, IMG_SIZE)

        if self.transform:
            img = self.transform(image=img)["image"]
            img = np.transpose(img, (2, 0, 1)).astype(np.float32)
            # img = np.expand_dims(img, 0).astype(np.float32)

            img_gray = self.transform(image=img_gray)["image"]
            img_gray = np.expand_dims(img_gray, 0).astype(np.float32)

        return {
            "image": img,
            "image_gray": img_gray,
            "stroke": stroke,
            "token": token,
            "targets": label
        }
```

The interesting is, when I change the code from
```python
drawing = self.drawings[idx]
```
to 

```python
drawing = self.drawings[0]
```
There is no more memory leak. I could not explain why because I still use this function on my own code written by myself and there is no leak.

Do you have any idea?

**Comment by buriy** [comment]
Hi,

First of all, are you sure it's a leak, not just memory fragmentation
and not too-eager Dataloader piling up a lot of data at the GPU to be
processed later?
A usual advice to distinguish between the two first is to add
 gc.collect() + torch.cuda.empty_cache()
and check if it helped.
For the last two, an idea is to reduce the number of worker threads
and see if that helps.

On Sun, Nov 4, 2018 at 6:04 PM Nguyen Xuan Bac <notifications@github.com> wrote:
>
> Updated:
>
> Train with bigger data by using finetune
> I downloaded dogscats dataset from fastai and tried to run finetune from your examples. There is no leak issue during training one epoch
> It seems that the leak comes from my code
>
> Debug my code
> I am able to figure out where the leak comes from. It occurs at my custom dataset.
>
>     def __getitem__(self, idx):
>
>         # Get token
>         token = self.tokens[idx]
>
>         # Get drawing
>         drawing = self.drawings[idx]
>         stroke = stroke_data(drawing).astype(np.float32)
>         # stroke = np.expand_dims(stroke, axis=0)
>         stroke = np.transpose(stroke, (1, 0)).astype(np.float32)
>
>         drawing = eval(drawing)
>         label = self.labels[idx]
>
>         # Plot the image
>         img, img_gray = drawing_to_image(drawing, IMG_SIZE, IMG_SIZE)
>
>         if self.transform:
>             img = self.transform(image=img)["image"]
>             img = np.transpose(img, (2, 0, 1)).astype(np.float32)
>             # img = np.expand_dims(img, 0).astype(np.float32)
>
>             img_gray = self.transform(image=img_gray)["image"]
>             img_gray = np.expand_dims(img_gray, 0).astype(np.float32)
>
>         return {
>             "image": img,
>             "image_gray": img_gray,
>             "stroke": stroke,
>             "token": token,
>             "targets": label
>         }
>
> The interesting is, when I change the code from
>
> drawing = self.drawings[idx]
>
> to
>
> drawing = self.drawings[0]
>
> There is no more memory leak. I could not explain why because I still use this function on my own code written by myself and there is no leak.
>
> Do you have any idea?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.



-- 
Best regards, Yuri V. Baburov, Skype: yuri.baburov

**Comment by ngxbac** [comment]
Hi,
Thank for your suggestion.
 

> For the last two, an idea is to reduce the number of worker threads
and see if that helps.

You are right. When I reduced the number of worker threads to 0. It helped. What is the problem here? 
Is there any better solution? Because running with workers=0 is very slow.

Many thanks

**Comment by buriy** [comment]
Try decreasing batch size then.

**Comment by Scitator** [comment]
Hi,
Sorry, I was out of time during last week.

I am not sure, but you can try things like:
- something like `drawing = self.drawings[idx].copy()`
- stop pinning memory in PyTorch DataLoader
- play with `n_workers` and `batch_size` parameters
- check @buriy idea and add `gc.collect() + torch.cuda.empty_cache()` after the end of each dataloder.

And it's always better if you can share the code (especially catalyst-based), so we can check out all parts of it.

PS. btw, @ngxbac feel free to PR your additional metrics & losses.

**Comment by ngxbac** [comment]
Hi,
Thank for your replies. 
None of the sugessions above could help me :D.
I searched in Pytorch forum. There are many people getting the similar issue. I guess, probably it comes from pytorch framework.
> PS. btw, @ngxbac feel free to PR your additional metrics & losses.

It is my pleasure to contribute to this awesome library. I will do it in the near future when I can clean my code, I promise.
